File: 39-combination-sum/39-combination-sum.java
class Solution {
    private void possibleComb(int candidates[],int target,List<List<Integer>> results,int index,ArrayList<Integer> comb,int sum){
        
        if(sum == target){
            results.add(new ArrayList<Integer>(comb));
            return;
        }
        if(index >= candidates.length) return;
        
        if(sum + candidates[index] <= target){
            comb.add(candidates[index]);
            possibleComb(candidates,target,results,index,comb,sum + candidates[index]);   
            comb.remove(comb.size() - 1);
        }
        
        possibleComb(candidates,target,results,index+1,comb,sum);
    }
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> results = new ArrayList<>();
        
        possibleComb(candidates,target,results,0,new ArrayList<Integer>(),0);
        return results;
    }
}

File: 39-combination-sum/39-combination-sum.cpp
class Solution {
public:
    void solve(vector<int> &arr,int idx,int target,vector<int> &path,vector<vector<int>> &result){
        if(target == 0){
            result.push_back(path);
            return;
        }
        if(idx >= arr.size() || target < 0) return;
        
        solve(arr,idx + 1,target,path,result);
        path.push_back(arr[idx]);
        
        solve(arr,idx,target - arr[idx],path,result);
        path.pop_back();
        
    }
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<int> path;
        vector<vector<int>> result;
        
        solve(candidates,0,target,path,result);
        return result;
    }
};